🔥 Use the `bulk_write` API to insert the incoming data instead of ins…

shankari opened the pull request:
…erting one by one

As we support larger deployments, the webapp is turning into a bottleneck. Looking at the logs, it looks like the bottleneck is again DocumentDB, which seems to take multiple milliseconds for each write, which adds up when we are saving many thousands of entries. In contrast, MongoDB inserts are < 1 ms each

MongoDB:

```
2025-04-16 07:27:58,093:INFO:140737036695104:Before modifying, keys_to_munge=[]
2025-04-16 07:27:58,093:INFO:140737036695104:(After modifying, entry_doc.keys()=dict_keys(['year', 'month', 'day', 'hour', 'minute', 'second', 'weekday', 'timezone'])
2025-04-16 07:27:58,093:INFO:140737036695104:Before modifying, keys_to_munge=[]
2025-04-16 07:27:58,093:INFO:140737036695104:(After modifying, entry_doc.keys()=dict_keys(['time_zone', 'plugin', 'write_ts', 'platform', 'read_ts', 'key', 'type', 'write_local_dt', 'write_fmt_time'])
2025-04-16 07:27:58,093:INFO:140737036695104:Before modifying, keys_to_munge=[]
2025-04-16 07:27:58,093:INFO:140737036695104:(After modifying, entry_doc.keys()=dict_keys(['year', 'month', 'day', 'hour', 'minute', 'second', 'weekday', 'timezone'])
2025-04-16 07:27:58,093:INFO:140737036695104:Before modifying, keys_to_munge=[]
2025-04-16 07:27:58,093:INFO:140737036695104:(After modifying, entry_doc.keys()=dict_keys(['battery_level_pct', 'battery_status', 'ts', 'local_dt', 'fmt_time'])
2025-04-16 07:27:58,093:INFO:140737036695104:Before modifying, keys_to_munge=[]
2025-04-16 07:27:58,093:INFO:140737036695104:(After modifying, entry_doc.keys()=dict_keys(['metadata', 'data', 'user_id'])
2025-04-16 07:27:58,095:DEBUG:140737036695104:Updated result for user = da63d4f7-ff86-4cca-bfc3-5fb60d81d5d3, key = background/battery, write_ts = 1722957271.6881871 = {'n': 1, 'nModified': 0, 'ok': 1.0, 'updatedExisting': True}

2025-04-16 07:27:58,095:INFO:140737036695104:Before modifying, keys_to_munge=[]
2025-04-16 07:27:58,095:INFO:140737036695104:(After modifying, entry_doc.keys()=dict_keys(['year', 'month', 'day', 'hour', 'minute', 'second', 'weekday', 'timezone'])
2025-04-16 07:27:58,095:INFO:140737036695104:Before modifying, keys_to_munge=[]
2025-04-16 07:27:58,095:INFO:140737036695104:(After modifying, entry_doc.keys()=dict_keys(['time_zone', 'plugin', 'write_ts', 'platform', 'read_ts', 'key', 'type', 'write_local_dt', 'write_fmt_time'])
2025-04-16 07:27:58,095:INFO:140737036695104:Before modifying, keys_to_munge=[]
2025-04-16 07:27:58,095:INFO:140737036695104:(After modifying, entry_doc.keys()=dict_keys(['year', 'month', 'day', 'hour', 'minute', 'second', 'weekday', 'timezone'])
2025-04-16 07:27:58,095:INFO:140737036695104:Before modifying, keys_to_munge=[]
2025-04-16 07:27:58,095:INFO:140737036695104:(After modifying, entry_doc.keys()=dict_keys(['battery_level_pct', 'battery_status', 'ts', 'local_dt', 'fmt_time'])
2025-04-16 07:27:58,095:INFO:140737036695104:Before modifying, keys_to_munge=[]
2025-04-16 07:27:58,095:INFO:140737036695104:(After modifying, entry_doc.keys()=dict_keys(['metadata', 'data', 'user_id'])
2025-04-16 07:27:58,095:DEBUG:140737036695104:Updated result for user = da63d4f7-ff86-4cca-bfc3-5fb60d81d5d3, key = background/battery, write_ts = 1722957271.8056412 = {'n': 1, 'nModified':
 0, 'ok': 1.0, 'updatedExisting': True}
```

DocumentDB:

```
2025-04-15T16:23:20.789-07:00 2025-04-15 23:23:20,789:DEBUG:140577872004672:Updated result for user = 6804aa5a-6fab-4a14-9aa5-8b0fb97f38cb, key = background/location, write_ts = 1744736224.032 = {'n': 1, 'nModified': 0, 'ok': 1.0, 'operationTime': Timestamp(1744759400, 1), 'updatedExisting': True}

2025-04-15T16:23:20.790-07:00 2025-04-15 23:23:20,789:INFO:140577872004672:Before modifying, keys_to_munge=[]
2025-04-15T16:23:20.790-07:00 2025-04-15 23:23:20,790:INFO:140577872004672:(After modifying, entry_doc.keys()=dict_keys(['key', 'platform', 'read_ts', 'time_zone', 'type', 'write_ts'])
2025-04-15T16:23:20.790-07:00 2025-04-15 23:23:20,790:INFO:140577872004672:Before modifying, keys_to_munge=[]
2025-04-15T16:23:20.790-07:00 2025-04-15 23:23:20,790:INFO:140577872004672:(After modifying, entry_doc.keys()=dict_keys(['accuracy', 'altitude', 'bearing', 'elapsedRealtimeNanos', 'filter', 'fmt_time', 'latitude', 'longitude', 'sensed_speed', 'ts'])
2025-04-15T16:23:20.790-07:00 2025-04-15 23:23:20,790:INFO:140577872004672:Before modifying, keys_to_munge=[]
2025-04-15T16:23:20.790-07:00 2025-04-15 23:23:20,790:INFO:140577872004672:(After modifying, entry_doc.keys()=dict_keys(['metadata', 'data', 'user_id'])

2025-04-15T16:23:20.797-07:00 2025-04-15 23:23:20,797:DEBUG:140577872004672:Updated result for user = 6804aa5a-6fab-4a14-9aa5-8b0fb97f38cb, key = background/filtered_location, write_ts = 1744736224.088 = {'n': 1, 'nModified': 0, 'ok': 1.0, 'operationTime': Timestamp(1744759400, 1), 'updatedExisting': True}

2025-04-15T16:23:20.798-07:00 2025-04-15 23:23:20,798:INFO:140577872004672:Before modifying, keys_to_munge=[]
2025-04-15T16:23:20.798-07:00 2025-04-15 23:23:20,798:INFO:140577872004672:(After modifying, entry_doc.keys()=dict_keys(['key', 'platform', 'read_ts', 'time_zone', 'type', 'write_ts'])
2025-04-15T16:23:20.799-07:00 2025-04-15 23:23:20,799:INFO:140577872004672:Before modifying, keys_to_munge=[]
2025-04-15T16:23:20.799-07:00 2025-04-15 23:23:20,799:INFO:140577872004672:(After modifying, entry_doc.keys()=dict_keys(['currState', 'transition', 'ts'])
2025-04-15T16:23:20.800-07:00 2025-04-15 23:23:20,800:INFO:140577872004672:Before modifying, keys_to_munge=[]
2025-04-15T16:23:20.800-07:00 2025-04-15 23:23:20,800:INFO:140577872004672:(After modifying, entry_doc.keys()=dict_keys(['metadata', 'data', 'user_id'])

2025-04-15T16:23:20.806-07:00 2025-04-15 23:23:20,806:DEBUG:140577872004672:Updated result for user = 6804aa5a-6fab-4a14-9aa5-8b0fb97f38cb, key = statemachine/transition, write_ts = 1744736224.159 = {'n': 1, 'nModified': 0, 'ok': 1.0, 'operationTime': Timestamp(1744759400, 1), 'updatedExisting': True}
```

This is also reflected in the overall time to save 10k entries.

DocumentDB:

```
2025-04-16T07:28:27.448-07:00 2025-04-16 14:28:27,448:DEBUG:140615831164480:START POST /usercache/put
2025-04-16T07:29:11.666-07:00 2025-04-16 14:29:11,666:DEBUG:140615831164480:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 44.25499749183655

2025-04-16T07:29:13.601-07:00 2025-04-16 14:29:13,601:DEBUG:140616266860096:START POST /usercache/put
2025-04-16T07:29:57.326-07:00 2025-04-16 14:29:57,326:DEBUG:140616266860096:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 43.75884938240051

2025-04-16T07:29:59.295-07:00 2025-04-16 14:29:59,295:DEBUG:140615873123904:START POST /usercache/put
2025-04-16T07:30:45.912-07:00 2025-04-16 14:30:45,912:DEBUG:140615873123904:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 46.65658187866211
```

MongoDB:

```
START 2025-04-16 14:58:54.065030 POST /usercache/put
END 2025-04-16 14:58:58.347547 POST /usercache/put da63d4f7-ff86-4cca-bfc3-5fb60d81d5d3 4.280893802642822
START 2025-04-16 14:58:59.265866 POST /usercache/put
END 2025-04-16 14:59:02.702028 POST /usercache/put da63d4f7-ff86-4cca-bfc3-5fb60d81d5d3 3.4349358081817627
START 2025-04-16 14:59:03.652562 POST /usercache/put
END 2025-04-16 14:59:07.195128 POST /usercache/put da63d4f7-ff86-4cca-bfc3-5fb60d81d5d3 3.5403733253479004
```

We are working on migrating from DocumentDB to mongoDB, but that is a major change and will require significant testing before we go live. However, in the short term, we can try to optimize the existing usage pattern of the database by using a `bulk_write` instead of multiple individual writes.

This is a much more targeted change, and it seems to give significant improvement, even for MongoDB. The overall time dropped from ~ 4 secs to ~ 1 second.

```
START 2025-04-16 15:04:03.372658 POST /usercache/put
END 2025-04-16 15:04:04.267284 POST /usercache/put da63d4f7-ff86-4cca-bfc3-5fb60d81d5d3 0.8934834003448486
START 2025-04-16 15:04:05.133535 POST /usercache/put
END 2025-04-16 15:04:06.171844 POST /usercache/put da63d4f7-ff86-4cca-bfc3-5fb60d81d5d3 1.035712480545044
START 2025-04-16 15:04:07.231908 POST /usercache/put
END 2025-04-16 15:04:08.166894 POST /usercache/put da63d4f7-ff86-4cca-bfc3-5fb60d81d5d3 0.9328651428222656
```

And the update was successful (this was re-running the same workload, so there were all matches, and nothing was inserted/upserted).

```
2025-04-16 15:04:07,609:INFO:140736934696512:For da63d4f7-ff86-4cca-bfc3-5fb60d81d5d3, about to perform bulk_write of 10000 operations from 2024-08-06T15:14:31.688187+00:00(1722957271.68818
71) -> 2024-08-07T00:01:23.792671+00:00(1722957271.6881871)
2025-04-16 15:04:08,160:DEBUG:140736934696512:Updated result for user = da63d4f7-ff86-4cca-bfc3-5fb60d81d5d3, is {'writeErrors': [], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 0
, 'nMatched': 10000, 'nModified': 0, 'nRemoved': 0, 'upserted': []}
2025-04-16 15:04:08,161:DEBUG:140736934696512:Updating user da63d4f7-ff86-4cca-bfc3-5fb60d81d5d3 with fields {'last_location_ts': 1723011084.8418388}
2025-04-16 15:04:08,161:DEBUG:140736934696512:User profile updated with data: {'last_location_ts': 1723011084.8418388}
```

Note that we are using an ordered `bulk_write`, which means that the write will fail as soon as any insert fails.
https://pymongo.readthedocs.io/en/stable/examples/bulk.html

> The first write failure that occurs (e.g. duplicate key error) aborts the remaining operations, and PyMongo raises BulkWriteError. The details attribute of the exception instance provides the execution results up until the failure occurred and details about the failure - including the operation that caused the failure.

This is actually consistent with the previous error handling where we were inserting one-by-one.
- If the result was not `ok`, we raised an exception
- If there was a `PyMongoError`, we also raised an exception

We would always stop processing the batch on the first error, so this change will not alter that behavior.

Actually inserting entries as opposed to just matching them is a bit slower but still much faster than the individual inserts.

```
START 2025-04-16 16:13:41.654027 POST /usercache/put
END 2025-04-16 16:13:43.653301 POST /usercache/put 872ea3ec-bbfb-42e8-8b9a-8675a5ed9e97 1.9958009719848633
START 2025-04-16 16:13:54.819003 POST /usercache/put
END 2025-04-16 16:13:55.992577 POST /usercache/put 872ea3ec-bbfb-42e8-8b9a-8675a5ed9e97 1.1658375263214111
START 2025-04-16 16:14:03.881300 POST /usercache/put
END 2025-04-16 16:14:04.868236 POST /usercache/put 872ea3ec-bbfb-42e8-8b9a-8675a5ed9e97 0.9839518070220947
```

```
2025-04-16 16:13:42,347:INFO:140737036695104:For 872ea3ec-bbfb-42e8-8b9a-8675a5ed9e97, about to perform bulk_write of 10000 operations from 2023-06-09T01:21:29.651001+00:00(1686273689.65100
1) -> 2023-10-11T16:11:43.098583+00:00(1686273689.651001)
2025-04-16 16:13:43,641:DEBUG:140737036695104:Updated result for user = 872ea3ec-bbfb-42e8-8b9a-8675a5ed9e97, is {'writeErrors': [], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 9
901, 'nMatched': 99, 'nModified': 0, 'nRemoved': 0, 'upserted': [{'index': 0, '_id': ObjectId('67ffd736f5b507db559db89d')}, {'index': 1, '_id': ObjectId('67ffd736f5b507db559db89e')}, {'index': 2, '_id': ObjectId('67ffd736f5b507db559db89f')}, {'index': 3, '_id': ObjectId('67ffd736f5b507db559db8a0')}, ...
{'index': 7797, '_id': ObjectId('67ffd737f5b507db559dd6de')}, {'index': 7798, '_id': ObjectId('67ffd737f5b507db559dd6df')}, {'index': 7799, '_id': ObjectId('67ffd737f5b507db559dd6e0')}, {'index': 7800, '_id': ObjectId('67ffd737f5b507db559dd6e1')}, {'index': 7801, '_id': ObjectId('67ffd737f5b507db559dd6e2')}, {'index': 7802, '_id': ObjectId('67ffd737f5b507db559dd6e3')}, {'index': 7803, '_id': ObjectId('67ffd737f5b507db559dd6e4')}, {'index': 7804, '_id': ObjectId('67ffd737f5b507db559dd6e5')}, {'index': 7805, '_id': ObjectId('67ffd737f5b507db559dd6e6')}, {'index': 7806, '_id': ObjectId('67ffd737f5b507db559dd6e7')}, {'index': 7807, '_id': ObjectId('67ffd737f5b507db559dd6e8')}, {'index': 7808, '_id': ObjectId('67ffd737f5b507db559dd6e9')}, {'index': 7809, '_id': ObjectId('67ffd737f5b507db559dd6ea')}, {'index': 7810, '_id': ObjectId('67ffd737f5b507db559dd6eb')}, {'index': 7811, '_id': ObjectId('67ffd737f5b507db559dd6ec')}, {'index': 7812, '_id': ObjectId('67ffd737f5b507db559dd6ed')}, {'index': 7813, '_id': ObjectId('67ffd737f5b507db559dd6ee')}, {'index': 7814, '_id': ObjectId('67ffd737f5b507db559dd6ef')}, {'index': 7815, '_id': ObjectId('67ffd737f5b507db559dd6f0')}, {'index': 7816, '_id': ObjectId('67ffd737f5b507db559dd6f1')}, {'index': 7817, '_id': ObjectId('67ffd737f5b507db559dd6f2')}, {'index': 7818, '_id': ObjectId('67ffd737f5b507db559dd6f3')}, {'index': 7819, '_id': ObjectId('67ffd737f5b507db559dd6f4')}, {'index': 7820, '_id': ObjectId('67ffd737f5b507db559dd6f5')}, {'index': 7821, '_id': ObjectId('67ffd737f5b507db559dd6f6')}, {'index': 7822, '_id': ObjectId('67ffd737f5b507db559dd6f7')}, {'index': 7823, '_id': ObjectId('67ffd737f5b507db559dd6f8')}, {'index': 7824, '_id': ObjectId('67ffd737f5b507db559dd6f9')}, {'index': 7825, '_id': ObjectId('67ffd737f5b507db559dd6fa')}, {'index': 7826, '_id': ObjectId('67ffd737f5b507db559dd6fb')}, {'index': 7827, '_id': ObjectId('67ffd737f5b507db559dd6fc')}, ...
{'index': 9990, '_id': ObjectId('67ffd737f5b507db559ddf40')}, {'index': 9991, '_id': ObjectId('67ffd737f5b507db559ddf41')}, {'index': 9992, '_id': ObjectId('67ffd737f5b507db559ddf42')}, {'index': 9993, '_id': ObjectId('67ffd737f5b507db559ddf43')}, {'index': 9994, '_id': ObjectId('67ffd737f5b507db559ddf44')}, {'index': 9995, '_id': ObjectId('67ffd737f5b507db559ddf45')}, {'index': 9996, '_id': ObjectId('67ffd737f5b507db559ddf46')}, {'index': 9997, '_id': ObjectId('67ffd737f5b507db559ddf47')}, {'index': 9998, '_id': ObjectId('67ffd737f5b507db559ddf48')}, {'index': 9999, '_id': ObjectId('67ffd737f5b507db559ddf49')}]}
```

shankari commented:
@JGreenlee can you be a second pair of eyes on this? I am going to test this on staging for a bit and try to get it to production today.

shankari commented:
@JGreenlee @rubinasingh @jgu2 for visibility

shankari commented:
I restarted the container, and then sent over everything from a dump of ccebikes, one by one.
1. The good news is that the memory utilization is fairly stable.
2. The bad news is that the CPU utilization is very bursty. I am not quite sure why, but for now, let's just verify that this is not very different than without this change.

<img width="1499" alt="Screenshot 2025-04-16 at 10 34 30 AM" src="https://github.com/user-attachments/assets/385f0fbb-ff7e-40c0-8ffa-8a47c5b802b3" />

JGreenlee reviewed (approved):
LGTM! Hopefully it helps significantly

shankari commented:
Before the changes, the usage was also spiky, but spikier

<img width="1499" alt="Screenshot 2025-04-16 at 12 45 49 PM" src="https://github.com/user-attachments/assets/4c11e9b2-8672-4701-b9d0-05a72e6f4f07" />

After the change, we had a few gaps where the CPU utilization was low, and the mean docker-level utilization seems to be around 50-60% instead of the current 80%

So this may not be a full solution, at least for MongoDB, but it seems like it will result in a ~ 10 percentage point difference, which is nothing to sneeze at.

Pushing this out to staging so we can see what it looks like in a DocDB environment

shankari commented:
Note also, for the record, that `update_many`, unlike `insert_many` does not support the shortcut of passing in a list of documents. It would need to pass in a list of (match, update) pairs, which is essentially going to be the same as passing a list of Upsert operations to `bulk_write`.

And we need to upsert so that we don't end up with duplicate entries if a push fails partway, so the phone re-pushes.

So we do have to use `bulk_write` instead of passing in a list as we do while bulk-saving in the analysis pipeline, for example.

shankari commented:
Re-ran the old code for exactly an hour to overlap with
https://github.com/e-mission/e-mission-server/pull/1047#issuecomment-2810260319

We can clearly see higher usage (fewer gaps) without the change

<img width="1499" alt="Screenshot 2025-04-16 at 2 00 56 PM" src="https://github.com/user-attachments/assets/99fa61fd-fbd5-47a1-89ed-7156128e5e75" />

shankari commented:
Running this on staging show some more marginal improvement

```
2025-04-16T15:25:29.920-07:00 2025-04-16 22:25:29,920:DEBUG:140632470701632:START POST /usercache/put
2025-04-16T15:26:08.110-07:00 2025-04-16 22:26:08,110:DEBUG:140632470701632:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 38.22941184043884 

2025-04-16T15:26:16.812-07:00 2025-04-16 22:26:16,812:DEBUG:140632144086592:START POST /usercache/put
2025-04-16T15:26:54.711-07:00 2025-04-16 22:26:54,711:DEBUG:140632144086592:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 37.94768571853638 

2025-04-16T15:27:03.322-07:00 2025-04-16 22:27:03,322:DEBUG:140631982728768:START POST /usercache/put
2025-04-16T15:27:37.918-07:00 2025-04-16 22:27:37,918:DEBUG:140631982728768:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 34.61320233345032 
```

So combined with https://github.com/e-mission/e-mission-server/pull/1046, we have improved by around 10-15 secs (~ 49 - ~ 35). It is a nice incremental improvement, but I don't think it is the order of magnitude improvement that we need.

shankari commented:
Ah but the CPU utilization is much much lower - it went from 8% to 5% to ~ 2%

<img width="1400" alt="Screenshot 2025-04-16 at 3 46 23 PM" src="https://github.com/user-attachments/assets/513f739d-a757-440b-ad3b-54227bbc8c25" />

shankari commented:
Couple of other quick checks before we move this over to production:
Q. what is 38 seconds actually spent on? Could it be that there are other bottlenecks than the database?
A: 37 of the 38 minutes in the first POST call are spent in the database

```
2025-04-16T15:25:29.920-07:00 2025-04-16 22:25:29,920:DEBUG:140632470701632:Called userCache.put
2025-04-16T15:25:30.124-07:00 2025-04-16 22:25:30,124:DEBUG:140632470701632:Old-style study, expecting token without a subgroup...
2025-04-16T15:25:30.229-07:00 2025-04-16 22:25:30,228:INFO:140632470701632:For 0e98a4f1-91ba-47a4-94a5-1f024fc5098f, about to perform bulk_write of 10000 operations from 2024-08-06T15:14:31.688187+00:00(1722957271.6881871) -> 2024-08-07T00:01:23.792671+00:00(1722957271.6881871)

2025-04-16T15:26:08.091-07:00 2025-04-16 22:26:08,091:DEBUG:140632470701632:Updated result for user = 0e98a4f1-91ba-47a4-94a5-1f024fc5098f, is {'writeErrors': [], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 0, 'nMatched': 10000, 'nModified': 0, 'nRemoved': 0, 'upserted': []}
2025-04-16T15:26:08.110-07:00 2025-04-16 22:26:08,110:DEBUG:140632470701632:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 38.22941184043884
```

I do notice that there is ~ 100 ms between receiving the call and checking the token; which is something we should take a look at next. But that pales in comparison to the 38 seconds it takes to contact the DB. 🤬 DocumentDB! MongoDB for an all match workload is 500 ms. We must switch to MongoDB ASAP; using DocDB was a big mistake in the first place.

shankari commented:
Q. What is the behavior when we actually insert entries instead of only upserting already existing entries?
A. 🤬 DocumentDB! Actually inserting entries instead of only matching them is a lot faster than individual upserts on mongodb. But on DocumentDB, if we push in new data, the response time goes back to 45 seconds

```
2025-04-16T16:24:54.144-07:00 2025-04-16 23:24:54,144:DEBUG:140632144086592:START POST /usercache/put
2025-04-16T16:25:39.366-07:00 2025-04-16 23:25:39,366:DEBUG:140632144086592:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 45.2470326423645
2025-04-16T16:25:43.909-07:00 2025-04-16 23:25:43,908:DEBUG:140631982728768:START POST /usercache/put
2025-04-16T16:26:28.390-07:00 2025-04-16 23:26:28,388:DEBUG:140631982728768:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 44.49021029472351
2025-04-16T16:26:32.889-07:00 2025-04-16 23:26:32,889:DEBUG:140631961749056:START POST /usercache/put
2025-04-16T16:27:16.028-07:00 2025-04-16 23:27:16,028:DEBUG:140631961749056:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 43.156694650650024
```

Although that doesn't even make sense. The first time we insert the data, it is 45 seconds. But the subsequent two times, we should have only had matches. Why are these also 45 seconds? The subsequent ones are in fact matches and are 45 seconds. I am not sure why this is - bigger objects? more recent objects?

```
2025-04-16T16:24:54.478-07:00 2025-04-16 23:24:54,478:INFO:140632144086592:For 0e98a4f1-91ba-47a4-94a5-1f024fc5098f, about to perform bulk_write of 10000 operations from 2023-06-09T01:21:29.651001+00:00(1686273689.651001) -> 2023-10-11T16:11:43.098583+00:00(1686273689.651001)
2025-04-16T16:25:39.338-07:00 2025-04-16 23:25:39,336:DEBUG:140632144086592:Updated result for user = 0e98a4f1-91ba-47a4-94a5-1f024fc5098f, is {'writeErrors': [], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 9901, 'nMatched': 99, 'nModified': 0, 'nRemoved': 0, 'upserted': [{'index': 0, '_id': ObjectId('68003c46610916c2a313d2ba')}, ...

2025-04-16T16:25:44.216-07:00 2025-04-16 23:25:44,215:INFO:140631982728768:For 0e98a4f1-91ba-47a4-94a5-1f024fc5098f, about to perform bulk_write of 10000 operations from 2023-06-09T01:21:29.651001+00:00(1686273689.651001) -> 2023-10-11T16:11:43.098583+00:00(1686273689.651001)
2025-04-16T16:26:28.329-07:00 2025-04-16 23:26:28,328:DEBUG:140631982728768:Updated result for user = 0e98a4f1-91ba-47a4-94a5-1f024fc5098f, is {'writeErrors': [], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 0, 'nMatched': 10000, 'nModified': 0, 'nRemoved': 0, 'upserted': []}

2025-04-16T16:26:33.085-07:00 2025-04-16 23:26:33,085:INFO:140631961749056:For 0e98a4f1-91ba-47a4-94a5-1f024fc5098f, about to perform bulk_write of 10000 operations from 2023-06-09T01:21:29.651001+00:00(1686273689.651001) -> 2023-10-11T16:11:43.098583+00:00(1686273689.651001)
2025-04-16T16:27:16.003-07:00 2025-04-16 23:27:16,003:DEBUG:140631961749056:Updated result for user = 0e98a4f1-91ba-47a4-94a5-1f024fc5098f, is {'writeErrors': [], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 0, 'nMatched': 10000, 'nModified': 0, 'nRemoved': 0, 'upserted': []}
```

However, the CPU utilization was super low; I guess the CPU load has moved to DocDB?!
<img width="1400" alt="Screenshot 2025-04-16 at 4 52 52 PM" src="https://github.com/user-attachments/assets/42d14436-8a83-419d-8c97-03dcf96bd7e1" />

Let's try one more time and then deploy to production

shankari commented:
Trying one more time, everything is another 10 seconds slower.

```
2025-04-16T16:54:57.129-07:00 2025-04-16 23:54:57,128:DEBUG:140631961749056:START POST /usercache/put
2025-04-16T16:55:58.040-07:00 2025-04-16 23:55:58,040:DEBUG:140631961749056:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 60.93902015686035
    
2025-04-16T16:56:01.910-07:00 2025-04-16 23:56:01,910:DEBUG:140631940769344:START POST /usercache/put
2025-04-16T16:56:57.324-07:00 2025-04-16 23:56:57,324:DEBUG:140631940769344:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 55.43309831619263
    
2025-04-16T16:57:01.059-07:00 2025-04-16 23:57:01,059:DEBUG:140632470701632:START POST /usercache/put
2025-04-16T16:57:53.908-07:00 2025-04-16 23:57:53,908:DEBUG:140632470701632:END POST /usercache/put 0e98a4f1-91ba-47a4-94a5-1f024fc5098f 52.861902952194214
```

I think it may be getting slower as the number of entries increases, which doesn't bode well for long-term DocDB scalability.
🤬 DocumentDB!

However, the CPU lowering does seem to be reliable, so let's move this to production and then focus on moving over to MongoDB ASAP

<img width="1400" alt="Screenshot 2025-04-16 at 5 15 04 PM" src="https://github.com/user-attachments/assets/c926f791-fd98-4589-ada7-a559c0d676c3" />
